### ISSUE TITLE
Investigate Copy Offload Create Request overhead

### ISSUE THREAD

**Issue by NateThornton** [NearNodeFlash/nnf-dm#88]
From @mcfadden8 
> It appears that there is a multi-second cost for each of the DataMovement::CreateRequest. 

Need someone to investigate the create request overhead Marty observed, and see if any improvements can be made. If necessary, a new Create Request List endpoint is an option to amortize any overhead.

**Comment by bdevcich-hpe** [comment]
@mcfadden8: Could you provide more details on the delays you've been seeing? I assume you are using the C library?

**Comment by mcfadden8** [comment]
This is using the C++ library (https://github.com/NearNodeFlash/nnf-dm/tree/master/daemons/compute/lib-cpp).

The test was run within a signle node allocation that was created as follows:

`flux mini alloc -q windom -N1 -t 180 --setattr=system.dw="#DW jobdw capacity=1TiB type=lustre name=scrcache"`

Our C++ test program did the following loop 5 times:

Loop 5 times:
1.  Create/Write/Close 512KiB file to location in `$DW_JOB_scrcache`
2. `nnfdm::CreateResponse cr(source_file, dest_file); client->Create(workflow, cr, cresponse)...`
3. Wait for `STATE_COMPLETED`/`STATUS_SUCCESS`

The destination file system is also a lustre file system.

I noticed that each iteration of the loop took between 6 to 10 seconds to complete.

**Comment by mcfadden8** [comment]
Currently, the SCR library calls on Copy Offload operation per file rather than a single directory copy.  This is because SCR cannot control the destination location of the file path.  SCR maintains a mapping between the source (cached file) and the destination (applications' notion of where the file should reside in the parallel file system).

I think that this constrain will cause several Copy Offload operations to queue up or potentially oversubscribe MPI resources.

Perhaps it would be best if the Copy Offload API could be updated to take a vector of source/destination pairs?  This might allow for one Copy Offload container for a set of files?

**Comment by bdevcich-hpe** [comment]
I believe the culprit is k8s client-side rate-limiting. @mcfadden8 Could you look at the nnf-dm log on the compute node and confirm that you have messages like this in there?

```
[root@rabbit-compute-5 ~]# journalctl -u nnf-dm | grep client-side
...
Mar 09 13:50:40 rabbit-compute-5 nnf-dm[170825]: I0309 13:50:40.879397  170825 request.go:682] Waited for 1.79205669s due to client-side throttling, not priority and fairness, request: GET:https://10.30.106.249:6443/apis/nnf.cray.hpe.com/v1alpha1/namespaces/rabbit-node-2/nnfdatamovements/nnf-dm-node-6pb6n
...
Mar 09 13:50:50 rabbit-compute-5 nnf-dm[170825]: I0309 13:50:50.880093  170825 request.go:682] Waited for 4.194927408s due to client-side throttling, not priority and fairness, request: GET:https://10.30.106.249:6443/apis/nnf.cray.hpe.com/v1alpha1/namespaces/rabbit-node-2/nnfdatamovements/nnf-dm-node-jd8wj
...
```
The default settings for a k8s clients is 5 Queries Per Second (QPS) with a Burst of 10 concurrent calls above the QPS setting. These are way too low for our purposes. To Create 1 DM request, we're querying k8s 3-4 times. So in your case, 5 DM CreateRequests would end up with at least 15-20 queries and rate-limiting would kick in.

Increasing these numbers to 50 and 100, respectively, has alleviated the rate-limiting that I've noticed when creating 50 DM requests. When testing with 100, I've noticed a few 1s waits.

I've chosen 50 and 100 because that's what the [new defaults are for Kubelet](https://github.com/kubernetes/kubernetes/pull/116121). I think it's a reasonable place to start and I plan on making these values configurable so they can be tweaked on a compute node.

Also note that in that PR, they mention that client-side rate-limiting is going away: 
> we believe that we're ready to start our journey towards getting rid of client-side rate-limiting. Kubelet is the best first candidate, because:

**Comment by bdevcich-hpe** [comment]
I tried 500 concurrent DM requests with the new settings (QPS: 50, Burst: 100) and it took 68 seconds to get all the DM requests created.

With the new PRs, these numbers can always be tweaked in the nnf-dm daemon running on the compute nodes. However, the obvious downside is that the higher these numbers, the more impact there is on the overall k8s performance while DM requests are being created. Just something to be aware of.

**Comment by behlendorf** [comment]
@bdevcich-hpe that's looking good.  When I checked our logs I didn't see any similar rate limiting, but I may have missed it.  We'll see about trying to reproduce it again this week and take a closer look.

**Comment by bdevcich-hpe** [comment]
#94 has been merged. Please re-open this issue if you are still experiencing problems after this change.

**Comment by mcfadden8** [comment]
@bdevcich-hpe and @behlendorf , I got around to testing with this today and I am unfortunately seeing similar slowness.  Here is what I am seeing:

With a single process, do an offload copy of a 8MB file 7 times:
```
8MBx1Rank:
Completed checkpoint 1:  5.248890 secs, 2.097159e+06 bytes, 3.995433e+05 bytes/sec
Completed checkpoint 2:  5.199897 secs, 2.097159e+06 bytes, 4.033078e+05 bytes/sec
Completed checkpoint 3:  5.319343 secs, 2.097159e+06 bytes, 3.942515e+05 bytes/sec
Completed checkpoint 4:  5.235890 secs, 2.097159e+06 bytes, 4.005353e+05 bytes/sec
Completed checkpoint 5:  5.177578 secs, 2.097159e+06 bytes, 4.050464e+05 bytes/sec
Completed checkpoint 6:  5.166768 secs, 2.097159e+06 bytes, 4.058938e+05 bytes/sec
Completed checkpoint 7:  5.197884 secs, 2.097159e+06 bytes, 4.034640e+05 bytes/sec
Completed checkpoint 8:  5.185767 secs, 2.097159e+06 bytes, 4.044067e+05 bytes/sec
Completed checkpoint 9:  5.641303 secs, 2.097159e+06 bytes, 3.717508e+05 bytes/sec
Completed checkpoint 10:  5.178383 secs, 2.097159e+06 bytes, 4.049834e+05 bytes/sec
Completed checkpoint 11:  5.243854 secs, 2.097159e+06 bytes, 3.999271e+05 bytes/sec
FileIO: Min    0.38 MB/s	Max    0.38 MB/s	Avg    0.38 MB/s	Agg    0.38 MB/s
```
When I tried writing a 2GB file with one process 7 times using the Offload Copy I saw:
```
2GBx1Rank
Completed checkpoint 1:  11.942310 secs, 2.147484e+09 bytes, 1.798215e+08 bytes/sec
Completed checkpoint 2:  11.762365 secs, 2.147484e+09 bytes, 1.825724e+08 bytes/sec
Completed checkpoint 3:  11.721051 secs, 2.147484e+09 bytes, 1.832160e+08 bytes/sec
Completed checkpoint 4:  9.298406 secs, 2.147484e+09 bytes, 2.309518e+08 bytes/sec
Completed checkpoint 5:  10.743153 secs, 2.147484e+09 bytes, 1.998932e+08 bytes/sec
Completed checkpoint 6:  11.483224 secs, 2.147484e+09 bytes, 1.870105e+08 bytes/sec
Completed checkpoint 7:  9.718531 secs, 2.147484e+09 bytes, 2.209679e+08 bytes/sec
Completed checkpoint 8:  12.150757 secs, 2.147484e+09 bytes, 1.767366e+08 bytes/sec
Completed checkpoint 9:  10.905237 secs, 2.147484e+09 bytes, 1.969222e+08 bytes/sec
Completed checkpoint 10:  10.178265 secs, 2.147484e+09 bytes, 2.109872e+08 bytes/sec
Completed checkpoint 11:  11.698361 secs, 2.147484e+09 bytes, 1.835713e+08 bytes/sec
FileIO: Min  186.76 MB/s	Max  186.76 MB/s	Avg  186.76 MB/s	Agg  186.76 MB/s
```
It gets significantly worse when I attempt to have 4 processes each writing a 2GB file 7 times.

**Comment by behlendorf** [comment]
Just for completeness the `nnf-datamovement-1.0~beta6-1.t4.x86_64` package was built from that same commit (https://github.com/NearNodeFlash/nnf-dm/commit/24fa88a34).

**Comment by behlendorf** [comment]
And the `journalctl -u nnf-dm` log from the time of the data movement.  It looks like nothing interesting was logged when it was running.  Is there any additional debugging it would be worth enabling?  Either in the daemon or on the Kubernetes side.

```
Mar 31 07:34:35 hetchy13 systemd[1]: Started Near-Node Flash (NNF) Data Movement Service.
Mar 31 07:34:35 hetchy13 nnf-dm[127323]: 2023/03/31 07:34:35 Using default rest configuration
Mar 31 07:34:36 hetchy13 nnf-dm[127323]: 2023/03/31 07:34:36 Version: v0.0.1-24fa88a34
Mar 31 07:34:36 hetchy13 nnf-dm[127323]: 2023/03/31 07:34:36 Start listening at  /var/run/nnf-dm.sock
```

**Comment by bdevcich-hpe** [comment]
@mcfadden8 Can you explain the times a bit more? Is this the time for the actual data movement or just for the Create Request? If it's the former, I suggest we open a new ticket since this one deals with the slowdown on the Create. It's another issue altogether if the data movement itself is slow.

What file systems are being used here? gfs2 to lustre? Also, could you provide more detail on the code that you are running? I'd like to understand what these numbers mean.

**Comment by mcfadden8** [comment]
For comparison, I ran a test that would use offload copy from the rabbit and inline copy from the compute node and see the following (some instrumentation was added to show the polling):
```
Nodes: 1, Type: Offload Copy from Rabbit, size: 2GB
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm::DeleteRequest::Delete(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
Completed checkpoint 1:  9.718456 secs, 2.147484e+09 bytes, 2.209696e+08 bytes/sec
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm::DeleteRequest::Delete(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
Completed checkpoint 2:  10.786464 secs, 2.147484e+09 bytes, 1.990906e+08 bytes/sec
FileIO: Min  189.87 MB/s	Max  189.87 MB/s	Avg  189.87 MB/s	Agg  189.87 MB/s

Nodes: 1, Type: Inline (direct) copy from compute node, size: 2GB
Completed checkpoint 1:  3.438703 secs, 2.147484e+09 bytes, 6.245040e+08 bytes/sec
Completed checkpoint 2:  4.058661 secs, 2.147484e+09 bytes, 5.291114e+08 bytes/sec
FileIO: Min  504.60 MB/s	Max  504.60 MB/s	Avg  504.60 MB/s	Agg  504.60 MB/s
```

**Comment by mcfadden8** [comment]
In order to rule out IO, I also ran a test with a small file (2MB) and saw the following:
```
Nodes: 1, Type: Offload Copy from Rabbit, size: 2MB
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
nnfdm_wait: nnfdm::DeleteRequest::Delete(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.1/rank_0.ckpt)
Completed checkpoint 1:  5.203656 secs, 2.097159e+06 bytes, 4.030164e+05 bytes/sec
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm_stat(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
nnfdm_wait: nnfdm::DeleteRequest::Delete(/mnt/nnf/4ed47a32-ee02-4f97-bb02-30805f21f5bf-0/scr/martymcf/scr.defjobid/scr.dataset.2/rank_0.ckpt)
Completed checkpoint 2:  5.306488 secs, 2.097159e+06 bytes, 3.952066e+05 bytes/sec
FileIO: Min    0.38 MB/s	Max    0.38 MB/s	Avg    0.38 MB/s	Agg    0.38 MB/s

Nodes: 1, Type: Inline (direct) copy from compute node, size: 2MB
Completed checkpoint 1:  0.537435 secs, 2.097159e+06 bytes, 3.902163e+06 bytes/sec
Completed checkpoint 2:  0.635166 secs, 2.097159e+06 bytes, 3.301751e+06 bytes/sec
FileIO: Min    3.15 MB/s	Max    3.15 MB/s	Avg    3.15 MB/s	Agg    3.15 MB/s
```

**Comment by mcfadden8** [comment]
> #94 has been merged. Please re-open this issue if you are still experiencing problems after this change.

@bdevcich-hpe - I do not see how to reopen this issue.  I think that I may have had a confusing typo in my original email.  I had said "It appears that there is a multi-second cost for each of the DataMovement::CreateRequest.".  What I meant to say was "It appears that there is a multi-second cost for completing an offload copy of each DataMovement::CreateRequest".

**Comment by bdevcich-hpe** [comment]
I'm looking into this, @mcfadden8

**Comment by bdevcich-hpe** [comment]
If I'm understanding your 2MB example correctly, it looks like you're doing a CreateRequest, StatusRequest(1s)x5, then a DeleteRequest(). To me, that means there's an issue with how long the Data Movement Request itself is taking rather than the Copy Offload API response times. It shouldn't be taking 5 seconds to transfer 2MB. You did mention lustre-to-lustre, so perhaps there some overhead there.

I think the first thing we need to do is understand how long it's taking to perform the data movement. I recommend removing your `DeleteRequest() `so that the `NnfDataMovement` resource stays on the k8s cluster. You can then query that resource and pull the `status.startTime` and `status.endTime`. I'd like to know what that duration is on your 2MB test. You'll have to access the k8s cluster manually - you won't be able to do this from the compute node (I don't think).

Another avenue to take is to fake out the data movement itself. This way, we can get the end-to-end response times and without the data movement portion. The `NnfDataMovement` resource will still be created in k8s, but the command will just return success immediately. The Data Movement command can be overridden via the `nnf-dm-config` `ConfigMap`, which is in the `nnf-dm-system` namespace. Comment out the existing `data.nnf-dm-config.yaml.profiles.default.command` and set it to something like `true` to fake it out. It might also be useful to keep the `mpirun` command in place and replace `dcp` with `true` in case `mpirun` is a culprit here (`mpirun --allow-run-as-root --hostfile $HOSTFILE true`).

To recap, re-run your tests and:
1. Remove the delete and collect the start/end time from the `NnfDataMovement` resource
2. Override Data Movement to complete immediately and see what your end-to-end times are

There's more the investigate on this, but I'll stop here b/c I think these are the most interesting places to investigate first.

Let's connect on slack and I can walk you through this if you need help.

**Comment by bdevcich-hpe** [comment]
For 1, you should also set the `maxWaitTime` to 0

**Comment by bdevcich-hpe** [comment]
Update: We're seeing this consistently when we do lustre to lustre data movement. In this case, the data movement operation is launched with `mpirun` on the k8s worker node via the nnf-dm-controller pod. It targets the nnf node(s) (nnf-dm-worker pod).

In this example, it takes 3.6s to launch the `hostname` command via `mpirun`. There is a visible pause in the output before the ssh warning.

```shell
root@nnf-dm-controller-manager-557cbd6c6-h64sw:/# date && time mpirun --allow-run-as-root -host 10-42-2-89.dm.nnf-dm-system hostname 2>&1 | ts '[%Y-%m-%d %H:%M:%S]'
Fri Jun  2 16:01:00 UTC 2023
[2023-06-02 16:01:03] Warning: Permanently added '[10-42-2-89.dm.nnf-dm-system]:2222,[10.42.2.89]:2222' (ECDSA) to the list of known hosts.
[2023-06-02 16:01:04] nnf-dm-worker-kzqxk

real	0m3.660s
user	0m0.098s
sys	0m0.305s
root@nnf-dm-controller-manager-557cbd6c6-h64sw:/#
```

However, if you do `ssh` manually, it's only taking 200ms:

```shell
root@nnf-dm-controller-manager-557cbd6c6-h64sw:/# date && time ssh 10-42-2-89.dm.nnf-dm-system hostname 2>&1 | ts '[%Y-%m-%d %H:%M:%S]'
Fri Jun  2 16:02:37 UTC 2023
[2023-06-02 16:02:37] Warning: Permanently added '[10-42-2-89.dm.nnf-dm-system]:2222,[10.42.2.89]:2222' (ECDSA) to the list of known hosts.
[2023-06-02 16:02:37] nnf-dm-worker-kzqxk

real	0m0.274s
user	0m0.053s
sys	0m0.014s
```
Considering that ssh is working fine, I suspect there is an issue with `mpirun` and how it's using ssh or perhaps some initialization before the ssh connection is established.

**Comment by bdevcich-hpe** [comment]
Thanks to @mcfadden8, we've discovered two things contributing to the overhead:

1. The `nnf-dm-controller-manager` pod has resource limits. This is directly affecting `mpirun`'s initialization. The `nnf-dm-worker` pods do not have resource limits and I have confirmed that there is no overhead when launching `mpirun` from a worker pod and targeting another.
2. This may or may not have an affect, but the open files limit on the system defaults to `1048576`. When choosing a smaller value, like `2048`, HPE sees a reduction in overhead on one of our test systems. LLNL does not see a reduction for this.

For 1, we will remove the resource limits (and do so for nnf-sos for consistency). This is something we can re-enable in the future if necessary.

For 2, we can set a lower open files limit in the nnf-dm `Dockerfile` or we can use the `nnf-dm-config `configmap to define a value and then use InitContainers to set the limit on pod startup. Both solutions would require the use of `/etc/security/limits.conf`.

**Comment by bdevcich-hpe** [comment]
For setting the number of open files, I've been having trouble getting this to stick inside of the container. kubernetes does not provide a way to do this for a container either. There's been a issue open for a long time: https://github.com/kubernetes/kubernetes/issues/3595.

However, one thing that does work is prefixing the data movement command with `ulimit -n 2048`:

```yaml
command: ulimit -n 2048 && ulimit -a && mpirun --allow-run-as-root --hostfile $HOSTFILE dcp --progress 1 --uid $UID     --gid $GID $SRC $DEST
```

Performing a lustre to lustre data movement (and storing the output), the open files limit has been set to 2048. 

```
Message:              real-time non-blocking time  (microseconds, -R) unlimited
core file size              (blocks, -c) unlimited
data seg size               (kbytes, -d) unlimited
scheduling priority                 (-e) 0
file size                   (blocks, -f) unlimited
pending signals                     (-i) 126237
max locked memory           (kbytes, -l) 64
max memory size             (kbytes, -m) unlimited
open files                          (-n) 2048
pipe size                (512 bytes, -p) 8
POSIX message queues         (bytes, -q) 819200
real-time priority                  (-r) 0
stack size                  (kbytes, -s) 8192
cpu time                   (seconds, -t) unlimited
max user processes                  (-u) unlimited
virtual memory              (kbytes, -v) unlimited
file locks                          (-x) unlimited
Warning: Permanently added '[10-42-3-89.dm.nnf-dm-system]:2222,[10.42.3.89]:2222' (ECDSA) to the list of known hosts.\r
[2023-06-06T20:23:26] Walking /mnt/nnf/3f232df9-ac98-4e54-96f2-116fa1d7b4e9-0/test
[2023-06-06T20:23:26] Walked 1 items in 0.001 secs (1900.826 items/sec) ...
[2023-06-06T20:23:26] Walked 1 items in 0.001 seconds (1587.864 items/sec)
[2023-06-06T20:23:26] Copying to /lus/global
[2023-06-06T20:23:26] Items: 1
[2023-06-06T20:23:26]   Directories: 0
[2023-06-06T20:23:26]   Files: 1
[2023-06-06T20:23:26]   Links: 0
[2023-06-06T20:23:26] Data: 0.000 B (0.000 B per file)
[2023-06-06T20:23:26] Creating 1 files.
```

With that, I don't think a fix is necessary in the container itself for setting the limit. Plus, this adds some flexibility and visibility. Do you want the limit to be part of the default data movement command, @mcfadden8?

**Comment by bdevcich-hpe** [comment]
PR: #113
